File: src/main/java/frc/utils/BatteryVoltageSim.java
package frc.utils;

import java.util.ArrayList;
import java.util.function.DoubleSupplier;

import edu.wpi.first.wpilibj.simulation.BatterySim;
import edu.wpi.first.wpilibj.simulation.RoboRioSim;

/*
 * singleton class to manage current draw for all simulated parts of the robot
 */
public class BatteryVoltageSim {
    private static BatteryVoltageSim instance;

    private double voltage = 11.5;//unloaded voltage
    private double nominalcurrent = 20.0;//default current draw
    private double current = 0.0;

    private ArrayList<DoubleSupplier> currentSources = new ArrayList<>();

    private BatteryVoltageSim(){
        currentSources.add(()->nominalcurrent);
    }
    public static synchronized BatteryVoltageSim getInstance(){
        if(instance == null){
            instance = new BatteryVoltageSim();
        }
        return instance;
    }

    public void addCurrentSource(DoubleSupplier currentSource){
        currentSources.add(currentSource);
    }
    
    public double calculateVoltage(){
        current = 0.0;
        for(DoubleSupplier currentSource : currentSources){
            current += currentSource.getAsDouble();
        }

        double Loadvoltage = BatterySim.calculateLoadedBatteryVoltage(voltage, 0.015, current);
        RoboRioSim.setVInVoltage(voltage);
        return Loadvoltage;
    }    

}

File: src/main/kotlin/frc/utils/BatteryVoltageSim.kt
package frc.utils

import edu.wpi.first.wpilibj.simulation.BatterySim
import edu.wpi.first.wpilibj.simulation.RoboRioSim
import java.util.function.DoubleSupplier

/**
 * singleton class to manage current draw for all simulated parts of the robot
 */
class BatteryVoltageSim private constructor() {

    private var voltage: Double = 11.5 // unloaded voltage
    private var nominalcurrent: Double = 20.0 // default current draw
    private var current: Double = 0.0

    private val currentSources: MutableList<DoubleSupplier> = ArrayList()

    init {
        currentSources.add(DoubleSupplier { nominalcurrent })
    }

    fun addCurrentSource(currentSource: DoubleSupplier) {
        currentSources.add(currentSource)
    }

    fun calculateVoltage(): Double {
        current = 0.0
        for (currentSource in currentSources) {
            current += currentSource.asDouble
        }
        val loadVoltage = BatterySim.calculateLoadedBatteryVoltage(voltage, 0.015, current)
        RoboRioSim.setVInVoltage(voltage)
        return loadVoltage
    }

    companion object {
        @Volatile
        private var instance: BatteryVoltageSim? = null

        @JvmStatic
        @Synchronized
        fun getInstance(): BatteryVoltageSim {
            if (instance == null) {
                instance = BatteryVoltageSim()
            }
            return instance!!
        }
    }
}
